### ISSUE TITLE
Pillow Version Discrepancy

### ISSUE THREAD

**Issue by I-Sam-I** [paulgavrikov/visualkeras#77]
Using PIL (pillow) version 10.4.0 causes the legend's text not to appear. However, downgrading to 9.4.0 works as intended.

**Comment by Soontosh** [comment]
Which version of visualkeras are you using?

**Comment by Soontosh** [comment]
@I-Sam-I Is the issue resolved?

**Comment by RobertNolet** [comment]
I had the same issue:
visualkeras 0.1.3
pillow 10.1.0

Downgrading pillow to 9.4 fixed it.

**Comment by Soontosh** [comment]
Thanks for letting me know. I'll take a look at this issue and add compatibility for pillow 9.4+, if possible.

**Comment by Soontosh** [comment]
@I-Sam-I @RobertNolet Thank you for reporting this issue. I am unable to replicate the issue on my machine, using both Pillo 10.1.0 and Pillow 10.4.0.

Could you provide code snippets, your environment, and a list of all your Python packages so I can try reproducing the issue? Thanks!

**Comment by RobertNolet** [comment]
name: tensorflow2.13
channels:
  - conda-forge
dependencies:
  - aom=3.6.1=hb765f3a_0
  - appnope=0.1.3=pyhd8ed1ab_0
  - argon2-cffi=23.1.0=pyhd8ed1ab_0
  - argon2-cffi-bindings=21.2.0=py310h2aa6e3c_4
  - asttokens=2.4.0=pyhd8ed1ab_0
  - attrs=23.1.0=pyh71513ae_1
  - backcall=0.2.0=pyh9f0ad1d_0
  - backports=1.0=pyhd8ed1ab_3
  - backports.functools_lru_cache=1.6.5=pyhd8ed1ab_0
  - beautifulsoup4=4.12.2=pyha770c72_0
  - bleach=6.1.0=pyhd8ed1ab_0
  - brotli=1.1.0=hb547adb_1
  - brotli-bin=1.1.0=hb547adb_1
  - bzip2=1.0.8=h3422bc3_4
  - ca-certificates=2024.7.4=hf0a4a13_0
  - cairo=1.18.0=hd1e100b_0
  - cffi=1.16.0=py310hdcd7c05_0
  - comm=0.1.4=pyhd8ed1ab_0
  - contourpy=1.1.1=py310h38f39d4_1
  - cycler=0.12.1=pyhd8ed1ab_0
  - dav1d=1.2.1=hb547adb_0
  - debugpy=1.8.0=py310h1253130_1
  - decorator=5.1.1=pyhd8ed1ab_0
  - defusedxml=0.7.1=pyhd8ed1ab_0
  - entrypoints=0.4=pyhd8ed1ab_0
  - exceptiongroup=1.1.3=pyhd8ed1ab_0
  - executing=1.2.0=pyhd8ed1ab_0
  - expat=2.5.0=hb7217d7_1
  - ffmpeg=5.1.2=gpl_h9861f02_112
  - font-ttf-dejavu-sans-mono=2.37=hab24e00_0
  - font-ttf-inconsolata=3.000=h77eed37_0
  - font-ttf-source-code-pro=2.038=h77eed37_0
  - font-ttf-ubuntu=0.83=hab24e00_0
  - fontconfig=2.14.2=h82840c6_0
  - fonts-conda-ecosystem=1=0
  - fonts-conda-forge=1=0
  - fonttools=4.43.1=py310hd125d64_0
  - freetype=2.12.1=hadb7bae_2
  - fribidi=1.0.10=h27ca646_0
  - future=0.18.3=pyhd8ed1ab_0
  - gettext=0.21.1=h0186832_0
  - gmp=6.2.1=h9f76cd9_0
  - gnutls=3.7.8=h9f1a10d_0
  - graphite2=1.3.13=h9f76cd9_1001
  - harfbuzz=8.2.1=hf1a6348_0
  - icu=73.2=hc8870d7_0
  - importlib-metadata=6.8.0=pyha770c72_0
  - importlib_metadata=6.8.0=hd8ed1ab_0
  - importlib_resources=6.1.0=pyhd8ed1ab_0
  - ipykernel=6.25.2=pyh1050b4e_0
  - ipython=8.16.1=pyh31c8845_0
  - ipython_genutils=0.2.0=py_1
  - ipywidgets=8.1.1=pyhd8ed1ab_0
  - jedi=0.19.1=pyhd8ed1ab_0
  - jinja2=3.1.2=pyhd8ed1ab_1
  - joblib=1.3.2=pyhd8ed1ab_0
  - jsonschema=4.19.1=pyhd8ed1ab_0
  - jsonschema-specifications=2023.7.1=pyhd8ed1ab_0
  - jupyter=1.0.0=pyhd8ed1ab_10
  - jupyter_client=8.4.0=pyhd8ed1ab_0
  - jupyter_console=6.6.3=pyhd8ed1ab_0
  - jupyter_contrib_core=0.4.0=pyhd8ed1ab_0
  - jupyter_contrib_nbextensions=0.7.0=pyhd8ed1ab_0
  - jupyter_core=5.4.0=py310hbe9552e_0
  - jupyter_highlight_selected_word=0.2.0=pyhd8ed1ab_1006
  - jupyter_latex_envs=1.4.6=pyhd8ed1ab_1002
  - jupyter_nbextensions_configurator=0.6.1=pyhd8ed1ab_0
  - jupyterlab_pygments=0.2.2=pyhd8ed1ab_0
  - jupyterlab_widgets=3.0.9=pyhd8ed1ab_0
  - kiwisolver=1.4.5=py310h38f39d4_1
  - lame=3.100=h1a8c8d9_1003
  - lcms2=2.16=ha0e7c42_0
  - lerc=4.0.0=h9a09cb3_0
  - libass=0.17.1=hf7da4fe_1
  - libblas=3.9.0=19_osxarm64_openblas
  - libbrotlicommon=1.1.0=hb547adb_1
  - libbrotlidec=1.1.0=hb547adb_1
  - libbrotlienc=1.1.0=hb547adb_1
  - libcblas=3.9.0=19_osxarm64_openblas
  - libcxx=16.0.6=h4653b0c_0
  - libdeflate=1.20=h93a5062_0
  - libexpat=2.5.0=hb7217d7_1
  - libffi=3.4.2=h3422bc3_5
  - libgfortran=5.0.0=13_2_0_hd922786_1
  - libgfortran5=13.2.0=hf226fd6_1
  - libglib=2.78.0=h24e9cb9_0
  - libiconv=1.17=he4db4b2_0
  - libidn2=2.3.4=h1a8c8d9_0
  - libjpeg-turbo=3.0.0=hb547adb_1
  - liblapack=3.9.0=19_osxarm64_openblas
  - libopenblas=0.3.24=openmp_hd76b1f2_0
  - libopus=1.3.1=h27ca646_1
  - libpng=1.6.39=h76d750c_0
  - libsodium=1.0.18=h27ca646_1
  - libsqlite=3.43.2=h091b4b1_0
  - libtasn1=4.19.0=h1a8c8d9_0
  - libtiff=4.6.0=h07db509_3
  - libunistring=0.9.10=h3422bc3_0
  - libvpx=1.13.1=hb765f3a_0
  - libwebp-base=1.3.2=hb547adb_0
  - libxcb=1.15=hf346824_0
  - libxml2=2.11.5=h25269f3_1
  - libxslt=1.1.37=h1728932_1
  - libzlib=1.2.13=h53f4e23_5
  - llvm-openmp=17.0.3=hcd81f8e_0
  - lxml=4.9.3=py310h78afa71_1
  - markupsafe=2.1.3=py310h2aa6e3c_1
  - matplotlib-base=3.8.0=py310h9d2df84_2
  - matplotlib-inline=0.1.6=pyhd8ed1ab_0
  - mistune=3.0.1=pyhd8ed1ab_0
  - munkres=1.1.4=pyh9f0ad1d_0
  - nbclient=0.8.0=pyhd8ed1ab_0
  - nbconvert=7.9.2=pyhd8ed1ab_0
  - nbconvert-core=7.9.2=pyhd8ed1ab_0
  - nbconvert-pandoc=7.9.2=pyhd8ed1ab_0
  - nbformat=5.9.2=pyhd8ed1ab_0
  - ncurses=6.4=h7ea286d_0
  - nest-asyncio=1.5.8=pyhd8ed1ab_0
  - nettle=3.8.1=h63371fa_1
  - notebook=6.4.12=pyha770c72_0
  - openh264=2.3.1=hb7217d7_2
  - openjpeg=2.5.0=h4c1507b_3
  - openssl=3.3.1=hfb2fe0b_2
  - p11-kit=0.24.1=h29577a5_0
  - packaging=23.2=pyhd8ed1ab_0
  - pandas=2.1.1=py310h5924a0a_1
  - pandoc=3.1.3=hce30654_0
  - pandocfilters=1.5.0=pyhd8ed1ab_0
  - parso=0.8.3=pyhd8ed1ab_0
  - patsy=0.5.3=pyhd8ed1ab_0
  - pcre2=10.40=hb34f9b4_0
  - pexpect=4.8.0=pyh1a96a4e_2
  - pickleshare=0.7.5=py_1003
  - pillow=10.1.0=py310hfae7ebd_0
  - pip=23.3=pyhd8ed1ab_0
  - pixman=0.42.2=h13dd4ca_0
  - pkgutil-resolve-name=1.3.10=pyhd8ed1ab_1
  - platformdirs=3.11.0=pyhd8ed1ab_0
  - prometheus_client=0.17.1=pyhd8ed1ab_0
  - prompt-toolkit=3.0.39=pyha770c72_0
  - prompt_toolkit=3.0.39=hd8ed1ab_0
  - psutil=5.9.5=py310h2aa6e3c_1
  - pthread-stubs=0.4=h27ca646_1001
  - ptyprocess=0.7.0=pyhd3deb0d_0
  - pure_eval=0.2.2=pyhd8ed1ab_0
  - pycparser=2.21=pyhd8ed1ab_0
  - pyglet=1.5.27=py310hbe9552e_4
  - pygments=2.16.1=pyhd8ed1ab_0
  - pyobjc-core=10.0=py310hd07e440_0
  - pyobjc-framework-cocoa=10.0=py310hd07e440_1
  - pyparsing=3.1.1=pyhd8ed1ab_0
  - python=3.10.12=h01493a6_0_cpython
  - python-dateutil=2.8.2=pyhd8ed1ab_0
  - python-fastjsonschema=2.18.1=pyhd8ed1ab_0
  - python-tzdata=2023.3=pyhd8ed1ab_0
  - python_abi=3.10=4_cp310
  - pytz=2023.3.post1=pyhd8ed1ab_0
  - pyyaml=6.0.1=py310h2aa6e3c_1
  - pyzmq=25.1.1=py310h30b7201_1
  - qtconsole-base=5.4.4=pyha770c72_0
  - qtpy=2.4.0=pyhd8ed1ab_0
  - readline=8.2=h92ec313_1
  - referencing=0.30.2=pyhd8ed1ab_0
  - rise=5.7.1=py310hbe9552e_2
  - rpds-py=0.10.6=py310hd442715_0
  - scikit-learn=1.3.1=py310h417b086_1
  - scipy=1.11.3=py310hd1cfc7d_1
  - seaborn=0.13.0=hd8ed1ab_0
  - seaborn-base=0.13.0=pyhd8ed1ab_0
  - send2trash=1.8.2=pyhd1c38e8_0
  - setuptools=68.2.2=pyhd8ed1ab_0
  - six=1.16.0=pyh6c4a22f_0
  - soupsieve=2.5=pyhd8ed1ab_1
  - stack_data=0.6.2=pyhd8ed1ab_0
  - statsmodels=0.14.0=py310h50ce23c_2
  - svt-av1=1.4.1=h7ea286d_0
  - terminado=0.17.1=pyhd1c38e8_0
  - threadpoolctl=3.2.0=pyha21a80b_0
  - tinycss2=1.2.1=pyhd8ed1ab_0
  - tk=8.6.13=hb31c410_0
  - tornado=6.3.3=py310h2aa6e3c_1
  - traitlets=5.9.0=pyhd8ed1ab_0
  - trimesh=4.0.1=pyhd8ed1ab_0
  - typing_extensions=4.8.0=pyha770c72_0
  - tzdata=2023c=h71feb2d_0
  - unicodedata2=15.1.0=py310h2aa6e3c_0
  - wcwidth=0.2.8=pyhd8ed1ab_0
  - webencodings=0.5.1=pyhd8ed1ab_2
  - wheel=0.41.2=pyhd8ed1ab_0
  - widgetsnbextension=4.0.9=pyhd8ed1ab_0
  - x264=1!164.3095=h57fd34a_2
  - x265=3.5=hbc6ce65_3
  - xorg-libxau=1.0.11=hb547adb_0
  - xorg-libxdmcp=1.1.3=h27ca646_0
  - xz=5.2.6=h57fd34a_0
  - yaml=0.2.5=h3422bc3_2
  - zeromq=4.3.5=h965bd2d_0
  - zipp=3.17.0=pyhd8ed1ab_0
  - zlib=1.2.13=h53f4e23_5
  - zstd=1.5.5=h4f39d0f_0
  - pip:
      - absl-py==2.0.0
      - aggdraw==1.3.18.post0
      - arviz==0.16.1
      - astunparse==1.6.3
      - cachecontrol==0.13.1
      - cachetools==5.3.1
      - certifi==2023.7.22
      - charset-normalizer==3.3.0
      - chex==0.1.7
      - cloudpickle==3.0.0
      - cryptography==41.0.7
      - distrax==0.1.3
      - dm-tree==0.1.8
      - farama-notifications==0.0.4
      - filelock==3.13.1
      - firebase-admin==6.3.0
      - flatbuffers==23.5.26
      - gast==0.4.0
      - gdown==4.7.1
      - google-api-core==2.15.0
      - google-api-python-client==2.110.0
      - google-auth==2.23.3
      - google-auth-httplib2==0.1.1
      - google-auth-oauthlib==1.0.0
      - google-cloud-core==2.4.1
      - google-cloud-firestore==2.13.1
      - google-cloud-storage==2.13.0
      - google-crc32c==1.5.0
      - google-pasta==0.2.0
      - google-resumable-media==2.6.0
      - googleapis-common-protos==1.62.0
      - grpcio==1.60.0
      - grpcio-status==1.60.0
      - gymnasium==0.29.1
      - h5netcdf==1.3.0
      - h5py==3.10.0
      - httplib2==0.22.0
      - idna==3.4
      - jax==0.4.21
      - jaxlib==0.4.21
      - jaxopt==0.8.2
      - keras==2.13.1
      - libclang==16.0.6
      - llvmlite==0.41.1
      - markdown==3.5
      - ml-dtypes==0.3.1
      - msgpack==1.0.7
      - numba==0.58.1
      - numpy==1.22.4
      - oauthlib==3.2.2
      - opt-einsum==3.3.0
      - probml-utils==0.1.dev116+g638ebb9
      - proto-plus==1.23.0
      - protobuf==4.24.4
      - pyasn1==0.5.0
      - pyasn1-modules==0.3.0
      - pyjwt==2.8.0
      - pynndescent==0.5.11
      - pysocks==1.7.1
      - python-graphviz==0.20.1
      - regex==2023.10.3
      - requests==2.31.0
      - requests-oauthlib==1.3.1
      - rsa==4.9
      - tensorboard==2.13.0
      - tensorboard-data-server==0.7.1
      - tensorflow==2.13.0
      - tensorflow-estimator==2.13.0
      - tensorflow-macos==2.13.0
      - tensorflow-metal==0.8.0
      - tensorflow-probability==0.23.0
      - termcolor==2.3.0
      - texsoup==0.3.1
      - toolz==0.12.0
      - tqdm==4.66.1
      - typing-extensions==4.5.0
      - umap-learn==0.5.5
      - uritemplate==4.1.1
      - urllib3==2.0.7
      - visualkeras==0.1.3
      - werkzeug==3.0.0
      - wrapt==1.15.0
      - xarray==2023.12.0
      - xarray-einstats==0.6.0
prefix: /Users/robertnolet/miniforge3/envs/tensorflow2.13


The following code is run in a Jupyter Notebook on a M2 MacBook Pro and results in an image with coloured boxes for the legend, but no legend text:

import visualkeras

from tensorflow import keras 
from tensorflow.keras import layers

inputs = keras.Input(shape=(28, 28, 1))
x = layers.Conv2D(filters=32, kernel_size=3, activation="relu")(inputs)
x = layers.MaxPooling2D(pool_size=2)(x)
x = layers.Conv2D(filters=64, kernel_size=3, activation="relu")(x)
x = layers.MaxPooling2D(pool_size=2)(x)
x = layers.Conv2D(filters=128, kernel_size=3, activation="relu")(x)
x = layers.Flatten()(x)
outputs = layers.Dense(10, activation="softmax")(x)
model = keras.Model(inputs=inputs, outputs=outputs)
    
visualkeras.layered_view(model, legend=True)

**Comment by Soontosh** [comment]
Thank you so much for this! I'll take a look and resolve this issue ASAP

**Comment by silverleaf21** [comment]
I had the same invisible text issues with pillow 10.4.0 and visualkeras 1.0.3.

For me the combination of pillow 11.0.0 and visualkeras 1.0.2 solved the issue.
(I also tried pillow 9.4 and 9.5 and got it to work, but I am unsure if that was with 1.0.2 or 1.0.3).

I have a linux based system, if necessary I can provide further details.

Pillow itself was not the issue, I got visible text on a plain image from executing a minimal example taken from pillow documentation.

**Comment by Soontosh** [comment]
@silverleaf21 Thanks for reporting this. I'll work on pushing a solution within the next few hours

**Comment by Soontosh** [comment]
@RobertNolet @I-Sam-I @silverleaf21 @juliengdt A fix has been pushed, thank you so much for your help! Will push Pypi update tomorrow once the changes have been fully tested

**Comment by Soontosh** [comment]
The bug has been squashed as part of the latest release. View at: https://pypi.org/project/visualkeras/0.1.4/

Closing this issue now
